thingsboard_gateway/gateway/duplicate_detector.py

```diff
@@ -0,0 +1,104 @@
+
+
+class DuplicateDetector:
+    def __init__(self, connectors_storage):
+        self._connectors_data = {}
+        self._connectors_storage = connectors_storage
+
+    def persist_latest_values(self):
+        # TODO: periodically save latest values if persistent storage is enabled
+        pass
+
+    def filter_data(self, connector_name, new_data):
+        in_data_filter_enabled = new_data.get(SEND_ON_CHANGE_PARAMETER)
+        if isinstance(in_data_filter_enabled, bool) and not in_data_filter_enabled:
+            return new_data
+
+        device_name = new_data[DEVICE_NAME_PARAMETER]
+        if not in_data_filter_enabled:
+            connector = self._connectors_storage.get_connector(connector_name)
+            if not connector or not connector.is_filtering_enable(device_name):
+                return new_data
+
+        to_send = {ATTRIBUTES_PARAMETER: [], TELEMETRY_PARAMETER: []}
+
+        remaining_attributes_count = 0
+        filtered_attributes_count = 0
+        for attribute in new_data[ATTRIBUTES_PARAMETER]:
+            for key, new_value in attribute.items():
+                if self._update_latest_attribute_value(connector_name, device_name, key, new_value):
+                    to_send[ATTRIBUTES_PARAMETER].append(attribute)
+                    remaining_attributes_count += 1
+                else:
+                    filtered_attributes_count += 1
+
+        remaining_telemetry_count = 0
+        filtered_telemetry_count = 0
+        for ts_kv_list in new_data[TELEMETRY_PARAMETER]:
+            ts_added = False
+            ts = ts_kv_list.get(TELEMETRY_TIMESTAMP_PARAMETER)
+            ts_values = {}
+            for key, new_value in ts_kv_list.get(TELEMETRY_VALUES_PARAMETER, ts_kv_list).items():
+                if self._update_latest_telemetry_value(connector_name, device_name, key, new_value):
+                    ts_values[key] = new_value
+                    ts_added = True
+                    remaining_telemetry_count += 1
+                else:
+                    filtered_telemetry_count += 1
+            if ts_added:
+                to_send[TELEMETRY_PARAMETER].append(
+                    {TELEMETRY_TIMESTAMP_PARAMETER: ts, TELEMETRY_VALUES_PARAMETER: ts_values} if ts else ts_values)
+
+        if remaining_attributes_count > 0 or remaining_telemetry_count > 0:
+            log.warning("[%s] '%s' changed attributes %d from %d, changed telemetry %d from %d",
+                        connector_name, device_name,
+                        remaining_attributes_count, remaining_attributes_count + filtered_attributes_count,
+                        remaining_telemetry_count, remaining_telemetry_count + filtered_telemetry_count)
+            to_send[DEVICE_NAME_PARAMETER] = device_name
+            to_send[DEVICE_TYPE_PARAMETER] = new_data[DEVICE_TYPE_PARAMETER]
+            return to_send
+
+        log.warning("[%s] '%s' device data has not been changed", connector_name, device_name)
```

Comment:
It will be better to use `log.info(...)`

Reply:
Sorry, copy/paste. Maybe `debug`?

Reply:
We get a lot of issues where users don't see if data sending to TB, so I think it is better to see why it is happening without turning on DEBUG level.

Reply:
ok